// @gltf-transform/core/dist/constants.js
/**
 * Current version of the package.
 * @hidden
 */
export const VERSION = `v${PACKAGE_VERSION}`;
/** @hidden */
export const NAME = '@gltf-transform/core';
/** @hidden */
export const GLB_BUFFER = '__glb.bin';
/** String IDs for core {@link Property} types. */
export var PropertyType;
(function (PropertyType) {
    PropertyType["ACCESSOR"] = "Accessor";
    PropertyType["ANIMATION"] = "Animation";
    PropertyType["ANIMATION_CHANNEL"] = "AnimationChannel";
    PropertyType["ANIMATION_SAMPLER"] = "AnimationSampler";
    PropertyType["BUFFER"] = "Buffer";
    PropertyType["CAMERA"] = "Camera";
    PropertyType["MATERIAL"] = "Material";
    PropertyType["MESH"] = "Mesh";
    PropertyType["PRIMITIVE"] = "Primitive";
    PropertyType["PRIMITIVE_TARGET"] = "PrimitiveTarget";
    PropertyType["NODE"] = "Node";
    PropertyType["ROOT"] = "Root";
    PropertyType["SCENE"] = "Scene";
    PropertyType["SKIN"] = "Skin";
    PropertyType["TEXTURE"] = "Texture";
    PropertyType["TEXTURE_INFO"] = "TextureInfo";
    PropertyType["TEXTURE_SAMPLER"] = "TextureSampler";
})(PropertyType || (PropertyType = {}));
